mne/decoding/xdawn.py

```diff
@@ -0,0 +1,206 @@
+        if None, the covariance is estimated from the epochs signal.
+    reg : float | str | None (default None)
+        if not None, allow regularization for covariance estimation
+        if float, shrinkage covariance is used (0 <= shrinkage <= 1).
+        if str, optimal shrinkage using Ledoit-Wolf Shrinkage ('ledoit_wolf')
+        or Oracle Approximating Shrinkage ('oas').
+
+    Attributes
+    ----------
+    filters_ : dict of ndarray
+        If fit, the Xdawn components used to decompose the data for each event
+        type, else empty.
+    patterns_ : dict of ndarray
+        If fit, the Xdawn patterns used to restore M/EEG signals for each event
+        type, else empty.
+
+
+    See Also
+    --------
+    mne.preprocessing.Xdawn
+    CSP
+
+    References
+    ----------
+    [1] Rivet, B., Souloumiac, A., Attina, V., & Gibert, G. (2009). xDAWN
+    algorithm to enhance evoked potentials: application to brain-computer
+    interface. Biomedical Engineering, IEEE Transactions on, 56(8), 2035-2043.
+
+    [2] Rivet, B., Cecotti, H., Souloumiac, A., Maby, E., & Mattout, J. (2011,
+    August). Theoretical analysis of xDAWN algorithm: application to an
+    efficient sensor selection in a P300 BCI. In Signal Processing Conference,
+    2011 19th European (pp. 1382-1386). IEEE.
+    """
+
+    def __init__(self, n_components=2, n_chan=None, signal_cov=None,
+                 reg=None):
+        """init xdawn."""
+        if n_chan is None:
+            raise ValueError("n_chan cannot be none. Please provide the "
+                             "number of channels")
+        self.n_chan = n_chan
+        super(XdawnTransformer, self).__init__(n_components, signal_cov, reg)
+
+    def fit(self, X, y):
+        """Fit Xdawn from epochs.
+
+        Parameters
+        ----------
+        X : ndarray, shape(n_channels, n_times * n_freq)
+            Data of epochs.
+        y : ndarray shape(n_samples,)
+            Target values.
+
+        Returns
+        -------
+        self : XdawnTransformer instance
+            The XdawnTransformer instance.
+        """
+        from sklearn.preprocessing import LabelEncoder
+
+        if X.ndim != 2 or not isinstance(X, np.ndarray):
+            raise ValueError("X should be 2 dimensional ndarray")
+        if not isinstance(y, np.ndarray):
+            raise ValueError("Labels must be numpy array")
+
+        epochs_data = X.reshape(X.shape[0], self.n_chan, X.shape[1] //
+                                self.n_chan)
+        # Extract signal covariance
+        self._get_signal_cov(epochs_data)
+
+        # estimates evoked covariance
+        evokeds = dict()
+        toeplitz = dict()
+        classes = LabelEncoder().fit(y).classes_
+        shape = epochs_data.shape
+        for eid in classes:
+            mean_data = np.mean(epochs_data[eid].reshape(-1, shape[1],
+                                shape[2]), axis=0)
+            evokeds[eid] = mean_data
+            toeplitz[eid] = 1.0
```

Comment:
IMO it would be cleaner to internally generate an arbitrary `events` var (with non overlapping events) in `XDawnTransformer` and let `_Xdawn` take care of all cases. 

This will make it cleaner in later development to handle overlapping events by `XDawnTransformer`.